lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -388,201 +378,529 @@ export class UserAgentApplication {
-                .then((idToken) => {
-                    this._silentLogin = false;
-                    this._logger.info("Unified cache call is successful");
-                    resolve(idToken);
-                }, (error) => {
-                    this._silentLogin = false;
-                    this._logger.error("Error occurred during unified cache ATS");
-                    this.loginPopupHelper(resolve, reject, scopes, extraQueryParameters);
-                });
-        }
-         else {
-            this.loginPopupHelper(resolve, reject, scopes, extraQueryParameters );
-        }
-      });
-  }
+      // Extract ADAL id_token if it exists
+      var idTokenObject;
+      idTokenObject = this.extractADALIdToken();
 
-  private loginPopupHelper( resolve: any , reject: any, scopes: Array<string>, extraQueryParameters?: string) {
-      //TODO why this is needed only for loginpopup
-      if (!scopes) {
-          scopes = [this.clientId];
+      // silent login if ADAL id_token is retrieved successfully - SSO
+      if (idTokenObject && !scopes) {
+        this._logger.info("ADAL's idToken exists. Extracting login information from ADAL's idToken ");
+        extraQueryParameters = Utils.constructUnifiedCacheExtraQueryParameter(idTokenObject, extraQueryParameters);
+
+        this._silentLogin = true;
+        this.acquireTokenSilent([this.clientId], this.authority, this.getUser(), extraQueryParameters)
+        .then((idToken) => {
+          this._silentLogin = false;
+          this._logger.info("Unified cache call is successful");
+          // TODO: Change resolve to return AuthResponse object
+          resolve(idToken);
+        }, (error) => {
+          this._silentLogin = false;
+          this._logger.error("Error occurred during unified cache ATS");
+          this.loginPopupHelper(resolve, reject, scopes, extraQueryParameters);
+        });
       }
-      const scope = scopes.join(" ").toLowerCase();
-      var popUpWindow = this.openWindow("about:blank", "_blank", 1, this, resolve, reject);
-      if (!popUpWindow) {
-          return;
+      // else proceed with login
+      else {
+        this.loginPopupHelper(resolve, reject, scopes, extraQueryParameters );
       }
+    });
+  }
+  
+  /**
+   * Helper function to loginPopup
+   *
+   * @hidden
+   * @param resolve
+   * @param reject
+   * @param scopes
+   * @param extraQueryParameters
+   */
+  private loginPopupHelper( resolve: any , reject: any, scopes: Array<string>, extraQueryParameters?: string) {
+    // TODO: why this is needed only for loginpopup
+    if (!scopes) {
+      scopes = [this.clientId];
+    }
+    const scope = scopes.join(" ").toLowerCase();
 
-      this._loginInProgress = true;
+    // Generate a popup window
+    // TODO: Refactor this so that openWindow throws an error, loginPopupHelper rejects or resolves based on that action
+    var popUpWindow = this.openWindow("about:blank", "_blank", 1, this, resolve, reject);
```

Comment:
if we are going to use promises, the right way to do them is chain actions 

this
.then
.then
.then
.catch 

so that every step of the way returns a promise and assumes success.  If at any point through there is and error, it can be handled in the catch.

passing around resolve and reject sets us up to not know the control flow of our app, and get into race conditions.,

Reply:
We are aware of this. But this refactor needs more consideration and touches the actual execution flow. I would recommend taking this up once the current PRs are merged and plan extensive testing around this. This is already a tracking AI internally.